### PR TITLE
feature : static SDK main entry withSentry.init(options),...

### PR DESCRIPTION
**Pull request by rxlabz**
…captureEvent(...), .captureException(...)

## :loudspeaker: Type of change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring

## :scroll: Description

Add an Sentry class with static methods to init the sdk and capture the events.

## :bulb: Motivation and Context

Sentry SDK unified API https://develop.sentry.dev/sdk/unified-api/#static-api

## :green_heart: How did you test it?

By checking if the internal client is correctly used when calling the static methods

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [ ] No breaking changes

## :crystal_ball: Next steps

more static methods

### CODE REVIEW

File: dart/example/main.dart [hunk 1]
```diff
@@ -5,7 +5,10 @@
 import 'dart:async';
 import 'dart:io';
 
-import 'package:sentry/sentry.dart';
+import 'package:sentry/src/sentry.dart';
+import 'package:sentry/src/sentry_options.dart';
```
Comment (marandaneto) [inline dart/example/main.dart]: ideally `sentry_options.dart` is already imported by `sentry.dart`

File: dart/lib/src/sentry.dart [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import 'stack_trace.dart';
+
+/// Sentry SDK main entry point
+///
+class Sentry {
+  static SentryClient _client;
+
+  Sentry._();
+
+  static void init(SentryOptions options) {
```
Comment (marandaneto) [inline dart/lib/src/sentry.dart]: instead of taking `SentryOptions` directly, if possible let's do it with a callback configurator, see:

https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/Sentry.java#L104-L111

https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/Sentry.java#L487-L500

so we have the change to instantiate `SentryOptions` ourselves and set default values, so people just mutate what they really need to afterwards

File: dart/lib/src/sentry.dart [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+      environmentAttributes: options.environmentAttributes,
+      compressPayload: options.compressPayload,
+      httpClient: options.httpClient,
+      clock: options.clock,
+      uuidGenerator: options.uuidGenerator,
+    );
+  }
+
+  /// Initializes the SDK
+  static void initDns(String dns) => _client = SentryClient(dsn: dns);
```
Comment (marandaneto) [inline dart/lib/src/sentry.dart]: we don't have this on other SDKs

File: dart/lib/src/sentry.dart [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  /// Reports an [event] to Sentry.io.
+  static Future<SentryResponse> captureEvent(
+    Event event, {
+    StackFrameFilter stackFrameFilter,
+  }) async {
+    return _client.captureEvent(
+      event: event,
+      stackFrameFilter: stackFrameFilter,
+    );
+  }
```
Comment (marandaneto) [inline dart/lib/src/sentry.dart]: our APIs return either a https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/protocol/SentryId.java
or an empty `SentryId` which is `00000000-0000-0000-0000-000000000000` I guess.
Also we never take a `stackFrameFilter` as a param, we can work on this stuff later, but this will need to be changed.

File: dart/lib/src/sentry_options.dart [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import 'utils.dart';
+
+/// Sentry SDK options
+class SentryOptions {
+  /// Default Log level if not specified Default is DEBUG
+  static final SeverityLevel defaultDiagnosticLevel = SeverityLevel.debug;
+
+  /// The DSN tells the SDK where to send the events to. If this value is not provided, the SDK will
+  ///  just not send any events.
+  final String dsn;
```
Comment (marandaneto) [inline dart/lib/src/sentry_options.dart]: if we decide to use the callback: https://github.com/getsentry/sentry-dart/pull/108/files#r503897735

we can't make them `final`

File: dart/lib/src/sentry_options.dart [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+  /// Contains [Event] attributes that are automatically mixed into all events
+  /// captured through this client.
+  ///
+  /// This event is designed to contain static values that do not change from
+  /// event to event, such as local operating system version, the version of
+  /// Dart/Flutter SDK, etc. These attributes have lower precedence than those
+  /// supplied in the even passed to [capture].
+  final Event environmentAttributes;
```
Comment (marandaneto) [inline dart/lib/src/sentry_options.dart]: ideally we'd follow already defined names https://github.com/getsentry/sentry-java/blob/main/sentry/src/main/java/io/sentry/SentryOptions.java#L117-L122
if they are about the same attribute

File: dart/pubspec.yaml [hunk 7]
```diff
@@ -6,7 +6,7 @@ description: >
 homepage: https://github.com/getsentry/sentry-dart
 
 environment:
-  sdk: ^2.0.0
+  sdk: ^2.1.0
```
Comment (marandaneto) [inline dart/pubspec.yaml]: why this change?

File: dart/test/sentry_test.dart [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+import 'package:test/test.dart';
+
+void main() {
+  group('Sentry.init', () {
+    SentryClient client;
+
+    Exception anException;
+
+    final dns =
+        'https://cb0fad6f5d4e42ebb9c956cb0463edc9@o447951.ingest.sentry.io/5428562';
```
Comment (marandaneto) [inline dart/test/sentry_test.dart]: we just need a parseable DSN here, but not a valid one, right?

File: dart/test/sentry_test.dart [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    test('should capture the exception', () {
+      Sentry.captureException(anException);
+      verify(client.captureException(exception: anException)).called(1);
+    });
+  });
+}
+
+class MockSentryClient extends Mock implements SentryClient {}
+
+final event = Event(
```
Comment (marandaneto) [inline dart/test/sentry_test.dart:39]: not possible to reuse the `event_example.dart ` right? just curious if there's some magic to share code across tests/samples

File: dart/lib/src/sentry_options.dart (hunk 6 above)
Comment (rxlabz) [reply dart/lib/src/sentry_options.dart]: renamed to `environment`

File: dart/test/sentry_test.dart (hunk 9 above)
Comment (rxlabz) [reply dart/test/sentry_test.dart:39]: not really as the test files are not exposed in the library.
We could set a relative path in the test file `import '../example/event_example.dart';` but I'm not sure I'll vote for this, do you ?

File: dart/lib/src/sentry_options.dart (hunk 5 above)
Comment (rxlabz) [reply dart/lib/src/sentry_options.dart]: yes indeed ! the class is now mutable.

File: dart/test/sentry_test.dart (hunk 9 above)
Comment (marandaneto) [reply dart/test/sentry_test.dart:39]: nah, not really

File: dart/lib/src/client.dart [hunk 10]
```diff
@@ -23,15 +23,15 @@ abstract class SentryClient {
...
+    Event environment,
     bool compressPayload,
     Client httpClient,
     dynamic clock,
     UuidGenerator uuidGenerator,
   }) =>
       createSentryClient(
         dsn: dsn,
-        environmentAttributes: environmentAttributes,
+        environmentAttributes: environment,
```
Comment (marandaneto) [inline dart/lib/src/client.dart]: I'd rename the named field as well

File: dart/example/main.dart [hunk 11]
```diff
@@ -77,3 +64,14 @@ Future<void> bar() async {
 Future<void> baz() async {
   throw StateError('This is a test error');
 }
+
+class Options implements OptionsConfiguration<SentryOptions> {
+  final String dsn;
+
+  Options({this.dsn});
```
Comment (marandaneto) [inline dart/example/main.dart]: why do we have the `dsn` here? I dont get it

File: dart/lib/src/sentry.dart (hunk 3 above)
Comment (rxlabz) [reply dart/lib/src/sentry.dart]: I removed it, but now the only way to init Sentry is to define an `OptionsConfiguration<SentryOptions>` which IMHO is not the simplest way for a beginner.

File: dart/lib/src/sentry.dart [hunk 12]
```diff
@@ -52,3 +47,9 @@ class Sentry {
   @visibleForTesting
   static void initClient(SentryClient client) => _client = client;
 }
+
+/// Configuration options callback
+abstract class OptionsConfiguration<T extends SentryOptions> {
```
Comment (marandaneto) [inline dart/lib/src/sentry.dart]: can't it be just an `interface`?

File: dart/example/main.dart [hunk 13]
```diff
@@ -77,3 +64,14 @@ Future<void> bar() async {
 Future<void> baz() async {
   throw StateError('This is a test error');
 }
+
+class Options implements OptionsConfiguration<SentryOptions> {
```
Comment (marandaneto) [inline dart/example/main.dart]: do we need this? the concept of lambda won't work the same as Java for example?

File: dart/lib/src/client.dart (hunk 10 above)
Comment (rxlabz) [reply dart/lib/src/client.dart]: 😳

File: dart/lib/src/sentry.dart (hunk 12 above)
Comment (rxlabz) [reply dart/lib/src/sentry.dart]: it's now a typedef / function signature

File: dart/lib/src/browser_client.dart [hunk 14]
```diff
@@ -17,15 +17,15 @@ import 'version.dart';
 
 SentryClient createSentryClient({
   @required String dsn,
-  Event environmentAttributes,
+  Event environment,
```
Comment (marandaneto) [inline dart/lib/src/browser_client.dart]: I'm a bit lost here, why an `Event` is called `environment`?

File: dart/lib/src/sentry.dart [hunk 15]
```diff
@@ -0,0 +1,54 @@
+import 'dart:async';
+
+import 'package:meta/meta.dart';
+
+import 'client.dart';
+import 'protocol.dart';
+import 'sentry_options.dart';
+
+/// Configuration options callback
+typedef OptionsConfiguration = void Function(SentryOptions);
```
Comment (marandaneto) [inline dart/lib/src/sentry.dart:10]: I liked that!

File: dart/example/main.dart (hunk 1 above)
Comment (marandaneto) [reply dart/example/main.dart]: @rxlabz any reason to ignore this one?

File: dart/lib/src/sentry.dart (hunk 3 above)
Comment (marandaneto) [reply dart/lib/src/sentry.dart]: our docs state that and all the SDKs should look like this, also the samples, it should not be an issue

File: dart/pubspec.yaml (hunk 7 above)
Comment (marandaneto) [reply dart/pubspec.yaml]: @rxlabz mind explaining?

File: dart/lib/src/browser_client.dart (hunk 14 above)
Comment (marandaneto) [reply dart/lib/src/browser_client.dart]: https://github.com/getsentry/sentry-dart/pull/108#discussion_r503902646

probably they were not exactly the same thing?

Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: from the doc : 
> /// [environment] contain event attributes that do not change over
  /// the course of a program's lifecycle. These attributes will be added to
  /// all events captured via this client. The following attributes often fall
  /// under this category: [Event.serverName], [Event.release], [Event.environment].

Comment (marandaneto) [reply dart/lib/src/browser_client.dart]: ok so then it makes sense to keep as `environmentAttributes` till we split them up to separate fields.

File: dart/pubspec.yaml (hunk 7 above)
Comment (rxlabz) [reply dart/pubspec.yaml]: an autofix options error : clicked on 'upgrade sdk' instead of 'import dart:async', and didn't realized

File: dart/lib/src/sentry_options.dart (hunk 6 above)
Comment (rxlabz) [reply dart/lib/src/sentry_options.dart]: reverted to `environmentAttributes`

File: dart/example/main.dart (hunk 1 above)
Comment (rxlabz) [reply dart/example/main.dart]: no it was updated

File: dart/lib/src/browser_client.dart (hunk 14 above)
Comment (rxlabz) [reply dart/lib/src/browser_client.dart]: ✅ reverted

File: dart/lib/src/sentry.dart (hunk 4 above)
Comment (rxlabz) [reply dart/lib/src/sentry.dart]: should I rename the `SentryResponse` class https://github.com/getsentry/sentry-dart/blob/f63ac278c1de1251a466920e0bd7d5c082477e34/dart/lib/src/client.dart#L234-L256 to `SentryId`?

File: dart/example/main.dart (hunk 1 above)
Comment (rxlabz) [reply dart/example/main.dart]: ah ok, sorry, I used to let the issue opener close it, but I'll close it when I'm sure it was resolved
.

File: dart/lib/src/sentry.dart (hunk 4 above)
Comment (marandaneto) [reply dart/lib/src/sentry.dart]: no, not really, this is the whole response, we'd need to create a `SentryId` class and inside of the client, we just read the `uuid` out of the response and create a `SentryId` with the proper `uuid`, this doesn't need to be in this PR

File: dart/example/main.dart (hunk 1 above)
Comment (marandaneto) [reply dart/example/main.dart]: no worries, I also close them when I see it, but if there are no comments or reactions, I'd need to look for the code, I think this one should have been closed by me btw haha
